pkg/ddc/alluxio/ufs_internal.go

```diff
@@ -201,32 +241,69 @@ func (e *AlluxioEngine) processUpdatingUFS(ufsToUpdate *utils.UFSToUpdate) (err
 		for _, mountRemove := range ufsToUpdate.ToRemove() {
 			err = fileUtils.UnMount(mountRemove)
 			if err != nil {
-				return err
+				return false, err
 			}
 		}
 	}
-	// need to reset ufsTotal to Calculating so that SyncMetadata will work
-	datasetToUpdate := dataset.DeepCopy()
-	datasetToUpdate.Status.UfsTotal = metadataSyncNotDoneMsg
-	if !reflect.DeepEqual(dataset.Status, datasetToUpdate.Status) {
-		err = e.Client.Status().Update(context.TODO(), datasetToUpdate)
+
+	return true, nil
+}
+
+// update alluxio mount using script in configmap
+func (e *AlluxioEngine) updateUFSWithMountConfigMapScript(dataset *datav1alpha1.Dataset) (updateReady bool, err error) {
+	// 1. update non native mount info according the data.Spec.Mounts
+	mountConfigMapName := e.getMountConfigmapName()
+	mountConfigMap, err := kubeclient.GetConfigmapByName(e.Client, mountConfigMapName, e.namespace)
+	if err != nil {
+		return false, err
+	}
+
+	nonNativeMounts, err := e.generateNonNativeMountsInfo(dataset)
+	if err != nil {
+		return false, err
+	}
+
+	newMountConfigMap := mountConfigMap.DeepCopy()
+	newMountConfigMap.Data[NON_NATIVE_MOUNT_DATA_NAME] = strings.Join(nonNativeMounts, "\n") + "\n"
+	if !reflect.DeepEqual(newMountConfigMap, mountConfigMap) {
+		e.Log.Info("update mount path", NON_NATIVE_MOUNT_DATA_NAME, nonNativeMounts)
+		err = kubeclient.UpdateConfigMap(e.Client, newMountConfigMap)
 		if err != nil {
-			e.Log.Error(err, "fail to update ufsTotal of dataset to Calculating")
+			return false, err
 		}
 	}
 
-	err = e.SyncMetadata()
+	// 2. execute mount script to mount and unmount alluxio path according to non native mount info
+	podName, containerName := e.getMasterPodInfo()
+	fileUtils := operations.NewAlluxioFileUtils(podName, containerName, e.namespace, e.Log)
+	err = fileUtils.ExecMountScripts()
 	if err != nil {
-		// just report this error and ignore it because SyncMetadata isn't on the critical path of Setup
-		e.Log.Error(err, "SyncMetadata", "dataset", e.name)
-		return nil
+		return false, errors.Wrapf(err, "execute mount.sh occurs error")
 	}
 
-	if everMounted {
-		e.updateMountTime()
+	// 3. compare alluxio mount paths and dataset mount path to mark updateReady because configmap updating has delay
+	// for alluxio master pods so above executing may not use the newest non native mount info.
+	mountedPaths, err := fileUtils.GetMountedAlluxioPaths()
+	// the root path can not be changed, skip it.
+	mountedPaths = utils.RemoveString(mountedPaths, "/")
+
+	var datasetMountPaths []string
+	for _, mount := range dataset.Spec.Mounts {
+		if common.IsFluidNativeScheme(mount.MountPoint) {
+			continue
+		}
+		m := utils.UFSPathBuilder{}.GenAlluxioMountPath(mount)
+		// skip root path mount
+		if m != "/" {
```

Comment:
Some times we also change the root path, should we skip it?

Reply:
In my opinion，Alluxio does not support change the root path.